File: server/src/application/entities/group.ts
import { randomUUID } from 'node:crypto';

import { Replace } from '@helpers/Replace';

import { Conversation } from './conversation';

interface GroupProps {
  name: string;
  description: string | null;
  createdAt: Date;
  updatedAt: Date | null;

  conversation: Conversation | null;
}

export class Group {
  private _id: string;
  private props: GroupProps;

  private constructor(props: GroupProps, id?: string) {
    this._id = id ?? randomUUID();

    this.props = {
      ...props,
      createdAt: props.createdAt || new Date(),
    };
  }

  public get id() {
    return this._id;
  }

  public get name() {
    return this.props.name;
  }

  public get description() {
    return this.props.description;
  }

  public get createdAt() {
    return this.props.createdAt;
  }

  public get conversation(): Conversation | null {
    return this.props.conversation;
  }

  public set conversation(conversation: Conversation) {
    this.props.conversation = conversation;
  }

  public get updatedAt() {
    return this.props.updatedAt;
  }

  static create(props: Replace<GroupProps, { createdAt?: Date; updatedAt?: Date }>, id?: string) {
    return new Group(
      {
        ...props,
        createdAt: props?.createdAt ?? new Date(),
        updatedAt: props?.updatedAt ?? null,
      },
      id ?? randomUUID(),
    );
  }
}

File: Proxymity-Chat-Service/src/ProxyMity.Domain/Entities/Group.cs
namespace ProxyMity.Domain.Entities;

[Table("group")]
public class Group
{
    [Key]
    public required Ulid Id { get; set; }

    [Required]
    public required string Name { get; set; }

    public string? Description { get; set; }

    [Required]
    public DateTime CreatedAt { get; set; }

    public DateTime? UpdatedAt { get; set; }

    [Required]
    public Ulid CreatedBy { get; set; }

    public Conversation Conversation { get; set; }

    public static Group Create(Ulid creatorBy, string groupName, string? groupDescription)
    {
        return new Group
        {
            Id = Ulid.NewUlid(),
            Name = groupName,
            Description = groupDescription,
            CreatedBy = creatorBy,
            CreatedAt = DateTime.UtcNow
        };
    }
}
